core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -766,7 +766,41 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
    * yielding the value of the first effect to succeed with a value.
    * Losers of the race will be interrupted immediately
    */
-  def raceAll[R1 <: R, E1 >: E, A1 >: A](ios: Iterable[ZIO[R1, E1, A1]]): ZIO[R1, E1, A1] = ZIO.raceAll(self, ios)
+  def raceAll[R1 <: R, E1 >: E, A1 >: A](ios: Iterable[ZIO[R1, E1, A1]]): ZIO[R1, E1, A1] = {
+    def arbiter[E1, A1](
+      fibers: Iterable[Fiber[E1, A1]],
+      done: Promise[E1, A1],
+      fails: Ref[Int],
+      idx: Int
+    )(res: Exit[E1, A1]): ZIO[R1, Nothing, _] =
+      res.foldM[R1, Nothing, Unit](
+        e =>
+          ZIO.flatten(fails.modify((c: Int) => (if (c == 0) done.complete(ZIO.halt(e)).unit else ZIO.unit) -> (c - 1))),
+        a =>
+          done
+            .complete(ZIO.succeed(a))
+            .flatMap(
+              set => if (set) Fiber.interruptAll(fibers.view.zipWithIndex.filter(_._2 != idx).map(_._1)) else ZIO.unit
+            )
+      )
+
+    (for {
+      done  <- Promise.make[E1, A1]
+      fails <- Ref.make[Int](ios.size)
+      c <- ZIO.uninterruptibleMask { restore =>
+            for {
+              head <- ZIO.interruptible(self).fork
+              tail <- ZIO.foreach(ios)(io => ZIO.interruptible(io).fork)
+              as   = head :: tail
+              _ <- as.view.zipWithIndex.foldLeft[ZIO[R1, E1, _]](ZIO.unit) {
```

Comment:
I think we could zipWithIndex just once and pass it to arbiter already zipped, WDYT?